server/recording_api.go

```diff
@@ -66,6 +55,156 @@ func (p *Plugin) recJobTimeoutChecker(callID, jobID string) {
 	}
 }
 
+func (p *Plugin) handleRecordingStartAction(state *channelState, callID, userID string) (*RecordingStateClient, httpResponse) {
+	var res httpResponse
+
+	if state.Call.Recording != nil && state.Call.Recording.EndAt == 0 {
+		res.Err = "recording already in progress"
+		res.Code = http.StatusForbidden
+		return nil, res
+	}
+
+	recState := new(recordingState)
+	recState.ID = model.NewId()
+	recState.CreatorID = userID
+	recState.InitAt = time.Now().UnixMilli()
+	state.Call.Recording = recState
+
+	if err := p.kvSetChannelState(callID, state); err != nil {
+		res.Err = fmt.Errorf("failed to set channel state: %w", err).Error()
+		res.Code = http.StatusInternalServerError
+		return nil, res
+	}
+
+	defer func() {
+		// In case of any error we relay it to the client.
+		if res.Err != "" {
+			recState.EndAt = time.Now().UnixMilli()
+			recState.Err = res.Err
+			p.publishWebSocketEvent(wsEventCallRecordingState, map[string]interface{}{
+				"callID":   callID,
+				"recState": recState.getClientState().toMap(),
+			}, &model.WebsocketBroadcast{ChannelId: callID, ReliableClusterSend: true})
+		}
+	}()
+
+	// Sending the event prior to making the API call to the job service
+	// since it could take a few seconds to complete and we want clients
+	// to get their local state updated as soon as it changes on the server.
+	p.publishWebSocketEvent(wsEventCallRecordingState, map[string]interface{}{
+		"callID":   callID,
+		"recState": recState.getClientState().toMap(),
+	}, &model.WebsocketBroadcast{ChannelId: callID, ReliableClusterSend: true})
+
+	// We don't want to keep the lock while making the API call to the service since it
+	// could take a while to return. We lock again as soon as this returns.
+	p.unlockCall(callID)
```

Comment:
We're unlocking a lock held by a caller? Yikes... 
If we intend this to be the official pattern, maybe we should be passing around a lock to make it explicit that it's allowed/expected?

Reply:
We are locking again below so it's not that bad. The tradeoff here is that I didn't want to repeat a bunch of state checking code in the function itself but if you think it's better to manage the lock completely from within this handler then I can revert.

If anything I think the stop case is much worse but we should be able to fix that soon since stopping a job won't require any API call once we merge the cloud-native stuff.

Reply:
Not advocating reverting, I just wanted to make sure we're balancing clarity with convenience. :)